dali/core/exec/tasking/scheduler.cc

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <cassert>
+#include <mutex>
+#include <iostream>
+#include "dali/core/exec/tasking/scheduler.h"
+
+namespace dali::tasking {
+
+bool Scheduler::CheckTaskReady(SharedTask &task) noexcept {
+  assert(task->state <= TaskState::Pending);
+
+  for (auto &w : task->preconditions_)
```

Comment:
Shouldn't we remove the `preconditions_` that are already meet so we don't have to check them again?

Reply:
We do remove completion events (see Notify) - other preconditions are all or none. This is an essential part of deadlock mitigation.

Reply:
Renamed the function.

Reply:
Exactly.